src/Interface/HasOrder.agda

```diff
@@ -0,0 +1,40 @@
+{-# OPTIONS --safe #-}
+
+module Interface.HasOrder where
+
+open import Level using (_⊔_; suc)
+open import Relation.Binary using (Rel)
+open import Relation.Binary.Definitions using (Decidable)
+open import Relation.Binary.Structures using (IsPreorder; IsPartialOrder; IsStrictTotalOrder)
+
+record HasPreorder {a ℓ ℓ₂} (A : Set a) (_≈_ : Rel A ℓ) : Set (a ⊔ ℓ ⊔ suc ℓ₂) where
```

Comment:
The problem is that now strict and non-strict orders are completely independent. If you want orders where the two can be defined independently, you need to relate them with a theorem. For example, you can use this:

```agda

record HasPreorder {a ℓ ℓ₂} (A : Set a) (_≈_ : Rel A ℓ) : Set (a ⊔ ℓ ⊔ suc ℓ₂) where
  infix 4 _≤_
  field
    _≤_ _<_    : Rel A ℓ₂
    isPreorder : IsPreorder _≈_ _≤_
    irrefl     : Irreflexive _≈_ _<_
    ≤⇔<∨≡      : ∀ {a b} → a ≤ b ⇔ (a < b ⊎ a ≈ b)

module _ {a ℓ} (A : Set a) (_≈_ : Rel A ℓ) (_≤_ : Rel A ℓ) (isPreorder : IsPreorder _≈_ _≤_)
         (_≈?_ : ∀ a b → (Dec (a ≈ b))) where
  open import Relation.Binary.Construct.NonStrictToStrict _≈_ _≤_

  hasPreorderFromNonStrict : HasPreorder A _≈_
  hasPreorderFromNonStrict = record
    { _≤_        = _≤_
    ; _<_        = _<_
    ; isPreorder = isPreorder
    ; irrefl     = <-irrefl
    ; ≤⇔<∨≡      = λ {a b} → mk⇔
      (λ a≤b → case a ≈? b of λ where (yes p) → inj₂ p ; (no ¬p) → inj₁ (a≤b , ¬p))
      λ where (inj₁ a<b) → proj₁ a<b ; (inj₂ a≈b) → IsPreorder.reflexive isPreorder a≈b
    }
```

This means it's easy to define `HasPreorder` instances because you can just use `hasPreorderFromNonStrict`, but if you really need separate definitions for the two orders you can supply them & prove that equivalence.

Reply:
[Here](https://input-output-rnd.slack.com/archives/C02HDH9BVU1/p1695865566318169?thread_ts=1695823291.359499&cid=C02HDH9BVU1) I explain why we should not add the `≤⇔<∨≡` field to the `HasPreorder` record (unless we believe there's something wrong with the definitions of strict and nonstrict orders on the `Epoch` type).

Reply:
This is a slack link, which isn't public, so there's no point in adding it here.

If there is some order where you can't make an instance of this record for that's fine, but in the example you've made on slack you can still make an instance by simply using `hasPreorderFromStrict`. It seems like the real problem you're having is that `EpochStructure` defines everything in terms of `_<ˢ_`, but there's no reason not to change that.

Reply:
Well, the slack link was for you... but okay, I'll avoid adding
non-public links to comments on GitHub in future.

Your suggestion is interesting... Correct me if I'm wrong, but I believe you're suggesting that I delete the existing definition of `_≤ᵉ_` and let it be defined by `hasPreorderFromStrict`, which will take care of proving `a ≤ᵉ b ⇔ (a <ᵉ b ⊎ a ≡ b)`.  However, I claim that for the Epoch type the "only if" direction of that equivalence is false... so something about this shouldn't work.

In any case, re. changing `EpochStructure` to not define everything in terms of `_<ˢ_`, I already did that in my latest commit (as I proposed to do in the comments on Slack).  I've rewritten everything in terms of `_≤ˢ_`  and it all works now except for one computation proof that I just commented out until I can get it working again.